luarules/gadgets/unit_smartfight.lua

```diff
@@ -0,0 +1,406 @@
+	local function GetClosestPointOnLine(Ax, Az, Bx, Bz, Px, Pz)
+		-- Nil Check inputs to be safe
+		if not Ax or not Az or not Bx or not Bz or not Px or not Pz then return 0,0,0 end
+		
+		local APx, APz = Px - Ax, Pz - Az
+		local ABx, ABz = Bx - Ax, Bz - Az
+		local ab2 = ABx*ABx + ABz*ABz
+		local ap_ab = APx*ABx + APz*ABz
+		if ab2 <= 0.001 then return Ax, Az, 0 end
+		local t = ap_ab / ab2
+		if t < 0 then t = 0 end
+		if t > 1 then t = 1 end
+		return Ax + ABx * t, Az + ABz * t, t
+	end
+
+	local function GetBestTarget(unitID)
+		local teamID = spGetUnitTeam(unitID)
+		local allyTeamID = spGetUnitAllyTeam(unitID)
+		local unitDefID = Spring.GetUnitDefID(unitID)
+		if not unitDefID then return nil end
+		
+		local unitDef = UnitDefs[unitDefID]
+		local unitStates = spGetUnitStates(unitID)
+		local moveState = unitStates.movestate
+		local searchRadius = 0
+		
+		if moveState == 2 then
+			searchRadius = 1000
+		elseif moveState == 1 then
+			searchRadius = (unitDef.maxWeaponRange) + unitDef.radius
+		elseif moveState == 0 then
+			return nil
+		end
+		
+		local x, _, z = spGetUnitPosition(unitID)
+		if not x then return nil end
+
+		local potentialTargets = Spring.GetUnitsInCylinder(x, z, searchRadius)
+		local bestTarget = nil
+		local bestDist = math.huge
+
+			for _, targetID in ipairs(potentialTargets) do
+			if spValidUnitID(targetID) then
+				
+				-- LOS CHECK: Is this unit visible to my ALLYTEAM? (Visual OR Radar)
+				local unitInLOS = spIsUnitInLos(targetID, allyTeamID)
+
+				if unitInLOS then
+					local capturingTeam = protectedTargets[targetID]
+					local isProtectedFromMe = (capturingTeam == teamID)
+					
+					if not isProtectedFromMe then
+						local targetTeam = spGetUnitTeam(targetID)
+						local isGaia = (targetTeam == gaiaTeamID)
+						local isNeutral = spGetUnitNeutral(targetID)
+						
+						if (not isGaia) and (not isNeutral) and (not Spring.AreTeamsAllied(teamID, targetTeam)) then
+							local dist = spGetUnitSeparation(unitID, targetID, true) or math.huge
+							if dist < bestDist then
+								bestDist = dist
+								bestTarget = targetID
+							end
+						end
+					end
+				end
+			end
+		end
+		return bestTarget
+	end
+
+	function gadget:CommandFallback(unitID, unitDefID, teamID, cmdID, cmdParams, cmdOptions)
+		if cmdID ~= CMD_SMARTFIGHT then return false end
+		if commandUpdateQueue[unitID] then return true, false end
+
+		if not cmdParams or #cmdParams < 3 then
+			if cmdParams and #cmdParams == 1 and spValidUnitID(cmdParams[1]) then
+				spGiveOrderToUnit(unitID, CMD_ATTACK,cmdParams[1])
+			end
+			return true, true
+		end
+
+		local targetX, targetY, targetZ = cmdParams[1], cmdParams[2], cmdParams[3]
+		local unitX, _, unitZ = spGetUnitPosition(unitID)
+		
+		-- Safety check: Unit might have died this frame
+		if not unitX then return true, false end
+
+		local distSq = (unitX - targetX)^2 + (unitZ - targetZ)^2
+		-- We've arrived at the destination, end the command
+		if distSq < 10000 then
```

Comment:
consider not using squares. also the 100 could be a configurable named constant
```suggestion
		local dist = math.diag(unitX - targetX, unitZ - targetZ)
		-- We've arrived at the destination, end the command
		if dist < 100 then
```